### ISSUE TITLE
How to use the CLI client to generate a schema?

### ISSUE THREAD

**Issue by djc** [tomhoule/graphql-client#83]
I've built a Juniper-based GraphQL server this week and now I'd like to execute tests against it. This project seems like the best way to do so from Rust, and so I'm trying to generate a schema for my server. The built-in GraphiQL instance can read the introspection data just fine, but when I run `graphql-client introspect-schema http://localhost:3000/graphql` I just get `{"data":{"schema":null},"errors":null}`.

(Just to be sure, I get the same result if I run `introspect-schema` against the GitHub GraphQL endpoint.)

What, if anything, am I getting wrong?

**Comment by tomhoule** [comment]
I haven't paid attention to the CLI since its initial release nearly two months ago so it may have broken. I will definitely investigate this asap but as a temporary solution you can use https://github.com/apollographql/apollo-cli for introspecting and downloading the schema.json

**Comment by djc** [comment]
I have been investigating, but can't quite figure it out. As far as I can tell the problem is in the deserialization of the introspection query result -- can you recommend any techniques for debugging the deserialization code that runs during `GraphQLResponse` deserialization?

**Comment by tomhoule** [comment]
The weird thing is that the CLI should panic in the event of a serde error, so the `schema` field in ResponseData was null (or absent). I don't remember exactly right now, but it might be `__schema` instead, I'll have time to investigate later today. If it happens with the github API as well I can probably reproduce the issue.

**Comment by djc** [comment]
Here's the response from my server: https://gist.githubusercontent.com/djc/c7de96fdb8e4e6839ec3f32045408a9f/raw/e86794a82cfd13cefd831bdb9686cbd0a1ee6830/introspection-result.json

**Comment by tomhoule** [comment]
Thanks! now think I know what the problem is: the root field of ResponseData should actually be `__schema` but we transform between snake and camel case, and the leading underscores most likely get swallowed. The CLI should be tested :/ I can't fix this right away, but I will do this evening - if you want to do it it's also welcome of course.

**Comment by djc** [comment]
So yeah, this change seems to at least get me some output:

```diff
diff --git a/graphql_query_derive/src/codegen.rs b/graphql_query_derive/src/codegen.rs
index f9501ca..34ece7a 100644
--- a/graphql_query_derive/src/codegen.rs
+++ b/graphql_query_derive/src/codegen.rs
@@ -120,7 +121,7 @@ pub(crate) fn response_for_query(
         #variables_struct
 
         #[derive(Debug, Serialize, Deserialize)]
-        #[serde(rename_all = "camelCase")]
+        //#[serde(rename_all = "camelCase")]
         pub struct ResponseData {
             #(#response_data_fields,)*
         }
diff --git a/graphql_query_derive/src/shared.rs b/graphql_query_derive/src/shared.rs
index 73332ff..e425920 100644
--- a/graphql_query_derive/src/shared.rs
+++ b/graphql_query_derive/src/shared.rs
@@ -20,7 +20,22 @@ pub(crate) fn render_object_field(
         };
     }
 
-    let name_ident = Ident::new(&field_name.to_snake_case(), Span::call_site());
+    let mut snake_case = field_name.to_snake_case();
+    let mut leading_underscores = 0;
+    for (i, c) in field_name.as_bytes().iter().enumerate() {
+        if *c != b'_' {
+            leading_underscores = i;
+            break;
+        }
+    }
+
+    if leading_underscores > 0 {
+        let mut new = field_name[..leading_underscores].to_string();
+        new.push_str(&snake_case);
+        snake_case = new;
+    }
+    let name_ident = Ident::new(&snake_case, Span::call_site());
 
     quote!(#description pub #name_ident: #field_type)
 }
```

But that means neither heck nor serde camel case conversions do what we need for GraphQL here...

**Comment by djc** [comment]
The schema JSON generated with that patch still caused problems with some missing fields.

This seems to be much more effective, though:

```diff
diff --git a/graphql_client_cli/src/main.rs b/graphql_client_cli/src/main.rs
index 5e5a387..253600d 100644
--- a/graphql_client_cli/src/main.rs
+++ b/graphql_client_cli/src/main.rs
@@ -85,10 +85,7 @@ fn introspect_schema(location: String, output: Option<PathBuf>) -> Result<(), fa
         println!("Something else happened. Status: {:?}", res.status());
     }
 
-    let json: graphql_client::GraphQLResponse<introspection_query::ResponseData> = res.json()?;
-    let json = serde_json::to_string(&json)?;
-
-    write!(out, "{}", json)?;
-
-    Ok(())
+    //let json: graphql_client::GraphQLResponse<introspection_query::ResponseData> = res.json()?;
+    let json: serde_json::Value = res.json()?;
+    Ok(serde_json::to_writer_pretty(out, &json)?)
 }
```

**Comment by tomhoule** [comment]
Oh indeed, that's a nice workaround.

**Comment by tomhoule** [comment]
I'm writing a test for the underscore issue, we can make a PR with your fix. I was planning to release today or just wait for the next nightly release (current one has a known bug that makes graphql-client's macros fail).

**Comment by tomhoule** [comment]
So the culprit for swallowing leading underscores is `#[serde(rename_all = "camelCase")]` apparently. For correctness it should rename to whatever the name of the field is in the schema. It shouldn't be too hard to correct, I'm on the bug in any case, thanks for discovering it, and sorry that you did.

**Comment by djc** [comment]
At this point I'm a bit wary of the round-trip through `ResponseData`, what is the point? Just taking the JSON result from the introspection query seems pretty good.

Note that I've tried to actually derive a simple mutation query, and I ran into trouble again. The `expand()` method mentioned in the README doesn't seem to get generated.

**Comment by tomhoule** [comment]
For now it's definitely a better option to just dump the whole response - if I remember correctly the idea was that we could have a nice error message if the response wasn't a valid schema, but it's supposing things are working. I'll look into the `expand` issue too.

**Comment by tomhoule** [comment]
Oh my, the README example is out of date... You can check out the api docs, they are tested: https://docs.rs/graphql_client/latest/graphql_client/trait.GraphQLQuery.html

**Comment by tomhoule** [comment]
#84 tries to address all the issues you encountered today, it's a bit rough around the edges, I'll get back to it a bit later and merge.

**Comment by tomhoule** [comment]
The issues raised here should all have been addressed by #84 so I'll close this for now, thanks a lot for the report and the assistance in finding a solution.
